### PR TITLE
Add a GetUp provider that will use the appropriate getup script

### PR DESCRIPTION
**Pull request by TrentHouliston**
This PR adds a Behaviour skill that when executed as a task will determine using `sensors.Htw` which way the robot is lying and then execute an appropriate script.

### CODE REVIEW

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Controls the minimum log level that NUClear log will display
+log_level: DEBUG
+
+scripts:
+  getup_upright:
+   - "Stand.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:6]: this seems strange

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  getup_left:
+   - "RollLeft.yaml"
+   - "GetUpFront.yaml"
+   - "Stand.yaml"
+  getup_right:
+   - "RollRight.yaml"
+   - "GetUpFront.yaml"
+   - "Stand.yaml"
+  getup_inverted:
+   - "Relax.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:22]: this also seems strange

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+log_level: DEBUG
+
+scripts:
+  getup_upright:
+   - "Stand.yaml"
+  getup_front:
+   - "GetUpFront.yaml"
+   - "Stand.yaml"
+  getup_back:
+   - "GetUpBack.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:11]: we can't get up from the back

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+   - "Stand.yaml"
+  getup_left:
+   - "RollLeft.yaml"
+   - "GetUpFront.yaml"
+   - "Stand.yaml"
+  getup_right:
+   - "RollRight.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:18]: these don't exist...

File: module/skill/GetUp/README.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Getup
+
+## Description
+
+A Behaviour module that will run the appropriate getup script for the
```
Comment (ysims) [inline module/skill/GetUp/README.md:5]: isn't this a skill module

**Review by ysims — COMMENTED** [review]
isn't is `skills` not `skill`

File: module/skill/GetUp/data/config/GetUp.yaml (hunk 3 above)
Comment (TrentHouliston) [reply module/skill/GetUp/data/config/GetUp.yaml:11]: Should change this config to do what you need to then :P

File: module/skill/GetUp/data/config/GetUp.yaml (hunk 4 above)
Comment (TrentHouliston) [reply module/skill/GetUp/data/config/GetUp.yaml:18]: I was making a guess

File: module/skill/GetUp/README.md (hunk 5 above)
Comment (TrentHouliston) [reply module/skill/GetUp/README.md:5]: It's using the behaviour system ¯\_(ツ)_/¯ I can change it if you want

File: module/skill/GetUp/data/config/GetUp.yaml (hunk 1 above)
Comment (TrentHouliston) [reply module/skill/GetUp/data/config/GetUp.yaml:6]: If you emit a getup task and the robot is upright I'd say the correct thing to do is to stand?
Any other getup script we run will result in the robot flinging itself to the ground.

Comment (ysims) [reply module/skill/GetUp/data/config/GetUp.yaml:6]: can we not just do nothing and emit done?

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Controls the minimum log level that NUClear log will display
+log_level: DEBUG
+
+scripts:
+  getup_upright:
+   - "Stand.yaml"
+  getup_front:
+   - "GetUpFront.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:8]:
```suggestion
   - "StandUpFront.yaml"
```

File: module/skill/GetUp/data/config/GetUp.yaml (hunk 3 above)
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:11]:
```suggestion
   - "RollOverBack.yaml"
   - "StandUpFront.yaml"
```

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+   - "Stand.yaml"
+  getup_back:
+   - "GetUpBack.yaml"
+   - "Stand.yaml"
+  getup_left:
+   - "RollLeft.yaml"
+   - "GetUpFront.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:15]:
```suggestion
   - "StandUpFront.yaml"
```

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+   - "Stand.yaml"
+  getup_right:
+   - "RollRight.yaml"
+   - "GetUpFront.yaml"
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:19]:
```suggestion
   - "StandUpFront.yaml"
```

File: module/skill/GetUp/src/GetUp.hpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            std::vector<::extension::behaviour::ScriptRequest> getup_front;
+            /// @brief Script sequence to run when getting from lying on the back to standing
+            std::vector<::extension::behaviour::ScriptRequest> getup_back;
+            /// @brief Script sequence to run when getting from lying on the right to standing
+            std::vector<::extension::behaviour::ScriptRequest> getup_right;
+            /// @brief Script sequence to run when getting from lying on the left to standing
+            std::vector<::extension::behaviour::ScriptRequest> getup_left;
+            /// @brief Script sequence to run when already upright
+            std::vector<::extension::behaviour::ScriptRequest> getup_upright;
+            /// @brief Script sequence to run when told to get up while upside down
```
Comment (ysims) [inline module/skill/GetUp/src/GetUp.hpp:25]: I think upside down is better than inverted. and you should explain why we relax when upside down.

**Review by ysims — APPROVED** [review]
sure

File: module/skill/GetUp/src/GetUp.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+                cfg.getup_upright.push_back(extension::behaviour::ScriptRequest(s));
+            }
+
+            cfg.getup_upside_down.clear();
+            for (const auto& s : config["scripts"]["getup_upside_down"].as<std::vector<std::string>>()) {
+                cfg.getup_upside_down.push_back(extension::behaviour::ScriptRequest(s));
+            }
+        });
+
+        on<Provide<GetUpTask>, Needs<BodySequence>, With<Sensors>>().then(
```
Comment (ysims) [inline module/skill/GetUp/src/GetUp.cpp:59]:
```suggestion
        on<Provide<GetUp>, Needs<BodySequence>, With<Sensors>>().then(
```

**Review by ysims — CHANGES_REQUESTED** [review]

File: module/skill/GetUp/src/GetUp.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+#include "message/skill/GetUp.hpp"
+
+namespace module::skill {
+
+    using extension::Configuration;
+    using extension::behaviour::Script;
+    using message::actuation::BodySequence;
+    using message::behaviour::state::Stability;
+    using message::input::Sensors;
+    using GetUpTask = message::skill::GetUp;
```
Comment (ysims) [inline module/skill/GetUp/src/GetUp.cpp:21]: whyyyy?

File: module/skill/GetUp/src/GetUp.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                        Eigen::Vector3d uXTw = Hwt.rotation().col(0);
+                        Eigen::Vector3d uYTw = Hwt.rotation().col(1);
+                        Eigen::Vector3d uZTw = Hwt.rotation().col(2);
+
+                        // Split into six cases (think faces of a cube) to work out which getup script we should run
+
+                        // torso x is mostly world z
+                        bool on_front = (uXTw.z() >= uXTw.x() && uXTw.z() >= uXTw.y());
+                        // torso x is mostly world -z
+                        bool on_back = (uXTw.z() <= uXTw.x() && uXTw.z() <= uXTw.y());
```
Comment (ysims) [inline module/skill/GetUp/src/GetUp.cpp:79]:
```suggestion
                        // torso x is mostly world -z
                        bool on_front = (uXTw.z() <= uXTw.x() && uXTw.z() <= uXTw.y());
                        // torso x is mostly world z
                        bool on_back = (uXTw.z() >= uXTw.x() && uXTw.z() >= uXTw.y());
```

File: module/skill/GetUp/data/config/GetUp.yaml [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# Controls the minimum log level that NUClear log will display
```
Comment (ysims) [inline module/skill/GetUp/data/config/GetUp.yaml:1]: Need a webots-specific without RollOverBack and just does StandUpBack

**Review by ysims — APPROVED** [review]

**Pull request merged**